x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    public static final class TypeParser extends AbstractGeometryFieldMapper.TypeParser {
+
+        @Override
+        @SuppressWarnings("rawtypes")
+        protected AbstractGeometryFieldMapper.Builder newBuilder(String name, Map<String, Object> params) {
+            if (params.containsKey(DEPRECATED_PARAMETERS_KEY)) {
+                return new LegacyGeoShapeFieldMapper.Builder(name,
+                    (LegacyGeoShapeFieldMapper.DeprecatedParameters)params.get(DEPRECATED_PARAMETERS_KEY));
+            }
+            return new GeoShapeWithDocValuesFieldMapper.Builder(name);
+        }
+
+        @Override
+        @SuppressWarnings("rawtypes")
+        public Mapper.Builder parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            Map<String, Object> params = new HashMap<>();
+            AbstractGeometryFieldMapper.Builder builder = (AbstractGeometryFieldMapper.Builder) super.parse(name, node, parserContext);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (TypeParsers.DOC_VALUES.equals(fieldName)) {
+                    params.put(TypeParsers.DOC_VALUES, XContentMapValues.nodeBooleanValue(fieldNode, name + "." + TypeParsers.DOC_VALUES));
+                    iterator.remove();
+                }
+            }
+
+            if (params.containsKey(TypeParsers.DOC_VALUES)) {
+                builder.docValues((Boolean) params.get(TypeParsers.DOC_VALUES));
+            }
+            return builder;
+        }
+    }
+
+    public GeoShapeWithDocValuesFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                                            Explicit<Boolean> ignoreMalformed, Explicit<Boolean> coerce,
+                                            Explicit<Boolean> ignoreZValue, Explicit<Boolean> docValues, Settings indexSettings,
+                                            MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, ignoreMalformed, coerce, ignoreZValue, indexSettings,
+            multiFields, copyTo);
+        this.docValues = docValues;
+    }
+
+    @Override
+    protected void doMerge(Mapper mergeWith) {
+        if (mergeWith instanceof LegacyGeoShapeFieldMapper) {
+            LegacyGeoShapeFieldMapper legacy = (LegacyGeoShapeFieldMapper) mergeWith;
+            throw new IllegalArgumentException("[" + fieldType().name() + "] with field mapper [" + fieldType().typeName() + "] " +
+                "using [BKD] strategy cannot be merged with " + "[" + legacy.fieldType().typeName() + "] with [" +
+                legacy.fieldType().strategy() + "] strategy");
+        }
+        super.doMerge(mergeWith);
```

Comment:
Can we add a test that defines the behaviour when merging mapping with different set up in doc values? (I guess this class does not exists together with `GeoShapeFieldMapper`? you have either one or the other)

Reply:
yeah, I am not sure what the best way of testing this is. I think what would be best is to have an integration test where there is a full-cluster restart switching from OSS to Default

Reply:
After speaking with the team, we came to the conclusion that since we do not have testing for it, we do not support flavor switches from one distribution to another. This is to mean that I should not need to support the scenario where one indexes documents with the geo-module GeoShapeFieldMapper.

To better understand the behavior, I did the following

1. start ES-OSS 8.0 and index one document with a geo_shape field (`POINT (-77.03653 38.897676)`)
2. restart cluster in ES-Default 8.0
3. index new document (`POINT(54.99, -10.89)`) with geo_shape and run a `geo_bounds` aggregation and a query.

Queries work well, the aggregation fails with the following exception:

```
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_state_exception",
                "reason": "unexpected docvalues type NONE for field 'location' (expected one of [BINARY, SORTED]). Re-index with correct docvalues type."
            }
        ],
        "type": "search_phase_execution_exception",
        "reason": "all shards failed",
        "phase": "query",
        "grouped": true,
        "failed_shards": [
            {
                "shard": 0,
                "index": "example",
                "node": "6ht3ufNbTR-mdcgSDu6ZGQ",
                "reason": {
                    "type": "illegal_state_exception",
                    "reason": "unexpected docvalues type NONE for field 'location' (expected one of [BINARY, SORTED]). Re-index with correct docvalues type."
                }
            }
        ],
        "caused_by": {
            "type": "illegal_state_exception",
            "reason": "unexpected docvalues type NONE for field 'location' (expected one of [BINARY, SORTED]). Re-index with correct docvalues type.",
            "caused_by": {
                "type": "illegal_state_exception",
                "reason": "unexpected docvalues type NONE for field 'location' (expected one of [BINARY, SORTED]). Re-index with correct docvalues type."
            }
        }
    },
    "status": 500
}
```

4. I force-merge the index and re-run the aggregation and the aggregation succeeds with an aggregation over the documents that were indexed with doc-values and ignores those that don't have doc values fields.

```
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 2,
            "relation": "eq"
        },
        "max_score": null,
        "hits": []
    },
    "aggregations": {
        "viewport": {
            "bounds": {
                "top_left": {
                    "lat": -10.897000003606081,
                    "lon": 54.999999990686774
                },
                "bottom_right": {
                    "lat": -10.897000003606081,
                    "lon": 54.999999990686774
                }
            }
        }
    }
}
```

So, from my understanding, the merging is successful, but aggregations over `geo_shape` fields who have whole segments without any doc-values will fail.

does this reflect your intuition?

Reply:
I've also changed the code such that there is more inherited from GeoShapeFieldMapper, so this copied merging code is no longer present

Reply:
umm, I was expecting the geo_shape field to have `docvalues == false` so even if you index one more point, it should not generate doc values. In that sense a user would need to reindex the data from OSS to Default distribution in order to have doc values.

Reply:
The intention is to have doc-values default to true. This presents a problem for existing indices since we cannot determine whether an index of a specific version was created using the OSS geo_shape field or the Default one.

That is why I asked whether we support the switching-flavors case. It is an inconvenient behavior that I'm happy to discuss further. It is just that a lack of the doc-values field implies that the default should be respected, which is `true`.

One solution is to not have an implicit doc-values setting, and always inject the parameter into a newly created mapping. that way if it is not present, it can be assumed to be `false`. It is slightly different from the existing behavior of Explicit settings.

Reply:
I discuss this with Tal offline and we agree this behaviour is acceptable as we do not really support moving from OSS to Default distribution. The recommendation to users would be to reindex in the case they want to run aggregations.